File: 2020/14.rb
def addressFloat(starting)
	frontier = [starting]
	while frontier.size > 0
		address = frontier.pop
		if address.count("X") == 0
			yield address.sub("X", "0")
			yield address.sub("X", "1")
		else
			frontier << address.sub("X", "0")
			frontier << address.sub("X", "1")
		end
	end
end

def store(lines, p2=false)
	memory = !p2 ? {}: []
	lines.each do |line|
		mask, numsTemp = line.split("\n", 2)
		nums = numsTemp.split("\n")	
		nums.each do |eq|
			key, value = eq.match(/\[(\d+)\] = (\d+)/).captures
			toBin = "%036b" % value.to_i
			keyToBin = "%036b" % key.to_i
			if !p2
				memory[key.to_i] = ""
			else
				memory << ["", toBin]
			end
			(0..mask.size-1).each do |bit|
				if !p2
					memory[key.to_i] << (mask[bit] == "X" ? toBin[bit]: mask[bit])
				else
					memory[-1][0] << (mask[bit] != "0" ? mask[bit]: keyToBin[bit])
				end
			end
		end
	end
	return memory
end

lines = File.open("14.txt").read.split("mask = ").map(&:strip).drop(1)
memory = store(lines)
puts memory.values.map {|v| v.to_i(2)}.reduce {|v, sum| sum + v}
memory = {}
adresses = store(lines, true)
adresses.map {|address| addressFloat(address[0]) {|newAddress| memory[newAddress] = address[1].to_i(2)}}
puts memory.values.sum

File: 2020/14.py
import re

lines = open("14.txt").read().split("\n")

# Apply the mask to a given value. `floating` applies to the second part of the puzzle and
# determines whether or not we return the string with X in it like in the second puzzle
def apply_mask(val, mask, floating=False):
    str_val = "{0:0{width}b}".format(val, width=len(mask))
    new_val = ""
    for d in range(len(mask)):
        # If we are not in part 2, we had the original value bit if the mask is `X`
        # If we are in part 2, we add the original value if the bit of the mask is `0`
        # Otherwise, we just add the value from the mask
        if (mask[d] == "X" and not floating) or (mask[d] == "0" and floating):
            new_val += str_val[d]
        else:
            new_val += mask[d]

    return int(new_val, 2) if not floating else new_val

def possible_addresses(masked_addr):
    possible_addresses = []
    x_count = masked_addr.count("X")

    # We will have 2^n different addresses where n is the number of floating bits in the mask
    for i in range(2**x_count):
        # We want to replace the different X's with every possible combination of bits.
        # Here we get a string of the bits of i and replace the Xs with that string in order and
        # add it to our possibles
        rep_string = '{0:0{width}b}'.format(i, width=x_count)
        new_addr = ""
        rep_idx = 0
        for d in range(len(mask)):
            if masked_addr[d] == "X":
                new_addr += rep_string[rep_idx]
                rep_idx += 1
            else:
                new_addr += masked_addr[d]
        possible_addresses.append(new_addr)

    return possible_addresses

memory = {}
mask = ""
for line in lines:
    if re.search('mask', line):
        mask = line.split(" = ")[1]
    else:
        m = re.match(r'mem\[(\d+)\] = (\d+)', line)
        addr, val = int(m.group(1)), int(m.group(2))
        memory[addr] = apply_mask(val, mask)

# Part 1: Went pretty quickly, no major issues
print (sum(memory.values()))

memory = {}
mask = ""
for line in lines:
    if re.search('mask', line):
        mask = line.split(" = ")[1]
    else:
        m = re.match(r'mem\[(\d+)\] = (\d+)', line)
        addr, val = int(m.group(1)), int(m.group(2))
        masked_addr = apply_mask(addr, mask, floating=True)
        for possible_addr in possible_addresses(masked_addr):
            memory[possible_addr] = val

# Part 2: Didn't realize that a `0` in the mask didn't overwrite the value to a zero, but rather 
# kept the bit the same. Spent some time on that.
print (sum(memory.values()))